### ISSUE TITLE
Graph data does not appear in PDF

### ISSUE THREAD

**Issue by cryptothiago** [IzakMarais/reporter#118]
I would like to understand why graphics data is not showing up in my PDF.

I installed all the modules. The pdf is generated, however the data of the graphs do not appear.

What is the possible problem?

![alt text](https://i.imgur.com/H4i0mQh.png)

**Comment by IzakMarais** [comment]
Can you provide the log output of the grafana-reporter? What version of grafana are you using? What datasource are you using?

**Comment by zacheryph** [comment]
I think I have tracked this down. I get the same issue with a dashboard that contains a single `Table` panel. After playing around with what reporter is requesting and what share is requesting it seems that the `d-solo` requests for tables do not support the relative time values?

* `?from=now%2FM&to=now` (this month so far) - _ERROR_
* `?from=now-30d&to=now` (last 30 days) - _ERROR_
* `?from=1554094800000&to=1554958816339` - WORKS!

I do not know how to work around this since the other panels seem to work just fine as reporter sends relative times for them, unless you were to send epoch for all? This might be something to raise up to Grafana.

**Comment by IzakMarais** [comment]
@zacheryph, maybe I am misunderstanding you, but if I submit the following to Grafana, it returns fine:

    http://localhost:3000/render/d-solo/Y4N5w6pmk/temp-test-text-panel?orgId=1&from=now-30d&to=now&panelId=5&width=1000&height=500&tz=Africa%2FJohannesburg

So there doesn't seem to be a problem on the Grafana side. 

I also generated reports by selecting `last 30 days` and `this month so far`  from the grafana dropdown and having grafana pass on these parameters to the reporter. In the reporter logs I could see it fetching the images with those parameters. All cases worked fine.

**Comment by zacheryph** [comment]
Apologize.  I am not sure what I was doing last night. Though it seems even more odd now. Here is what I am looking at now.  Reporter says success on all 4 but just relative shows the image as per description.  The simplest way around this is to add `refresh=1` to link params to hit reporter, that way you dont have to dig up the `orgId`.

* using epoch works
* using relative is broken
  * `_?from=now-30d&to=now&panelId=2&width=1000&height=500`
* relative with orgId works (and _fixes_ reporter)
  * `_?orgId=1&from=now-30d&to=now&panelId=2&width=1000&height=500`
* using relative with refresh defined works (and _fixes_ reporter)
  * `_?refresh=1m&from=now-30d&to=now&panelId=2&width=1000&height=500`

**Comment by IzakMarais** [comment]
Thanks for the further feedback. It does seem like a Grafana problem, not a reporter problem. However it is very strange that I cannot reproduce it.  I have tried with both grafana 4.1.2 (running on a remote server) and 6.0.2 (running on local host) and neither of them had problems using relative links for the d-solo endpoint. 

Can you try running grafana on your local host and seeing if you can still reproduce the behaviour? Are you seeing anything suspicious in the grafana logs? Alternatively, can you try a different datasource (what type are you using currently?).

**Comment by zacheryph** [comment]
This is a Postgresql data source. I only figured this out manually editing the URLs between what reporter was originally sending and what the Share link was generating. I do think its weird that either orgId or refresh need to be defined :-/

**Comment by IzakMarais** [comment]
Aah ok, so I removed the orgId paramter (copied your link) and tested again. The following still works the first time:

    http://localhost:3000/render/d-solo/Y4N5w6pmk/temp-test-text-panel?from=now-30d&to=now&panelId=5&width=1000&height=500

**Comment by IzakMarais** [comment]
Since we are seeing different results, I think (if you are willing to do go to the effort) it might be useful to eliminate other variables. See my suggestions above:

> Can you try running grafana on your local host and seeing if you can still reproduce the behaviour? Are you seeing anything suspicious in the grafana logs? Alternatively, can you try a different datasource (what type are you using currently?).

**Comment by zacheryph** [comment]
I'll see what I can do time allowing.  I currently only have postgresql as a source. We're using Grafana more for reporting currently than monitoring

**Comment by zacheryph** [comment]
OK, this is just even more weird now and have reason to believe its resource related? After deploying this to DigitalOcean and just refreshing the same report I can have differing results. succeeding, then failing. On report with multiple panels, having some succeed, refresh, others succeed. All succeed at one point and all fail at one point.

I will attempt to deploy this to a larger DigitalOcean instance this weekend and see if that resolves the issue.

**Comment by zacheryph** [comment]
I have had this running on a 16GB / 8 vCPU CPU Optimized instance on Digital Ocean and I still get the occasional error red flag. It actually runs more stable on the 4G/2vCPU CPU Optimized instance :-/

Not sure what else to do here

**Comment by IzakMarais** [comment]
If you feel up to it you could try modifying the source code to use less concurrency. At this line: https://github.com/IzakMarais/reporter/blob/master/report/report.go#L139 you could change the number of workers from 5 to 1 to see if that improves your reliability?

**Comment by zacheryph** [comment]
Yeah, I already tried that. I think the issues get less frequent the lower the number but even when I set workers to 1 i still get red flags.  I did discover the following however. Apparently Grafana, at least according to the log, is sending a `400` bad for some requests for some reason? Not sure if we could differentiate a _bad_ request vs a "bad" request ;). Maybe simply retrying 2-3 times (as needed) for reports that return this would work. I will look into this a little more if and/or when I get more time.

#### Successful Panel Grafana Log
```
t=2019-04-30T12:29:15+0000 lvl=info msg=Rendering logger=rendering path="d-solo/5KZ_k7Qik/_?..."
```
#### Failed Panel Grafana Log
```
t=...DATE... lvl=info msg=Rendering logger=rendering path="d-solo/5KZ_k7Qik/_?...REDACTED..."
t=...DATE... lvl=info msg="Request Completed" logger=... userId=0 orgId=1 uname= method=POST path=/api/tsdb/query status=400 remote_addr=127.0.0.1 time_ms=51 size=1537 referer="http://localhost:3000/d-solo/5KZ_k7Qik/_?..."
```

**Comment by IzakMarais** [comment]
Interestingly that log suggests that the tsdb is returning a 400 in response to grafana's request for time series data. Perhaps you could look in the db logs for clues

**Comment by zacheryph** [comment]
Knock on wood. I think upgrading to Grafana 6.1 resolved this. I am unable to reproduce after upgrading.

**Comment by omnia888** [comment]
Dear Izak ,
Dose IzakMarais reporter support grafana version 6.6.1
